datanommer.models/datanommer/models/__init__.py

```diff
@@ -128,7 +130,16 @@ def add(envelope):
     obj.msg = message['msg']
     obj.headers = headers
 
-    session.add(obj)
+    # Construct an INSERT ... SELECT ... which will do nothing if an
+    # entry with a matching msg_id already exists. This can happen in the
+    # event of message redelivery.
+    cols = [c for c in Message.__table__.c if c.name != 'id']
+    vals = [literal(getattr(obj, c.name)) for c in cols]
+    not_exists = ~exists(select([Message.__table__.c.i]).\
+                         where(Message.msg_id == msg_id))
+    insert = Message.__table__.insert().from_select(cols, select(vals).\
+                                                    where(not_exists))
+    session.execute(insert)
```

Comment:
There are two cases here. One is duplicate message delivery. This should be a rare/exceptional case, but appears to happen in the case of failover/reconnection. Agreed that this *could* be handled by a try/catch, I just chose to do it this way instead. If it's possible to avoid triggering an error condition, it seems like it's sensible to do so.

The other case is two workers processing two different messages, but which both relate to the same user or package. The get_or_create() method had a race condition between checking for the existence of that user/package and inserting it. When there was a collision, the second worker would hit a unique constraint violation, and the message would get lost. Handling this case requires either nested transactions, or the INSERT ... SELECT construct used in Singleton.get_or_create().